### PR TITLE
feat: Support Top level workflow template reference 

### PR DESCRIPTION
**Pull request by sarabala1979**
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [ ] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [ ] I've signed the CLA.
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to USERS.md.

### CODE REVIEW

File: examples/workflow-template/top-level-tmpl-ref.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  generateName: workflow-template-hello-world-
+spec:
+  entrypoint: whalesay-template
+  arguments:
+    parameters:
+      - name: message
+        value: "hello world"
+  workflowTemplateRef:
+    name: workflow-template-whalesay-template
+
```
Comment (jessesuen) [inline examples/workflow-template/top-level-tmpl-ref.yaml:13]: Can we have two examples? One that utilizes arguments/entrypoint and one that doesn't. 

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: workflow-template-hello-world-
spec:
  workflowTemplateRef:
    name: workflow-template-whalesay-template
```

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 2]
```diff
@@ -809,6 +812,22 @@ type TemplateRef struct {
 	ClusterScope bool `json:"clusterScope,omitempty" protobuf:"varint,4,opt,name=clusterScope"`
 }
 
+// WorkflowTemplateRef is a reference of workflow template resource on Top level workflow.
```
Comment (jessesuen) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:815]: Why is `Top` capitalized?

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 3]
```diff
@@ -809,6 +812,22 @@ type TemplateRef struct {
...
+type WorkflowTemplateRef struct {
+	// Name is the resource name of the workflow template.
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+	// ClusterScope indicates the referred template is cluster scoped (i.e., a ClusterWorkflowTemplate).
+	ClusterScope bool `json:"clusterscope,omitempty" protobuf:"varint,2,opt,name=clusterscope"`
```
Comment (jessesuen) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:820]: Isn't this supposed to be camel cased?

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 4]
```diff
@@ -903,6 +922,9 @@ type WorkflowStatus struct {
 
 	// ResourcesDuration is the total for the workflow
 	ResourcesDuration ResourcesDuration `json:"resourcesDuration,omitempty" protobuf:"bytes,12,opt,name=resourcesDuration"`
+
+	//StoredWorkflowTemplateSpec stores the top level WorkflowTemplate spec
```
Comment (jessesuen) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:926]: Space after `//`

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 5]
```diff
@@ -268,6 +268,9 @@ type WorkflowSpec struct {
 
 	// Shutdown will shutdown the workflow according to its ShutdownStrategy
 	Shutdown ShutdownStrategy `json:"shutdown,omitempty" protobuf:"bytes,33,opt,name=shutdown,casttype=ShutdownStrategy"`
+
+	// workflowTemplateRef holds WorkflowTemplate reference -TODO-Bala update comments
```
Comment (jessesuen) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:272]: You have a TODO here.

File: workflow/common/convert.go [hunk 6]
```diff
@@ -26,16 +26,21 @@ func ConvertCronWorkflowToWorkflow(cronWf *wfv1.CronWorkflow) *wfv1.Workflow {
...
-func ConvertWorkflowTemplateToWorkflow(template *wfv1.WorkflowTemplate) *wfv1.Workflow {
-	wf := toWorkflow(template.TypeMeta, template.ObjectMeta, template.Spec.WorkflowSpec)
-	wf.Labels[LabelKeyWorkflowTemplate] = template.ObjectMeta.Name
-	return wf
-}
-
-func ConvertClusterWorkflowTemplateToWorkflow(template *wfv1.ClusterWorkflowTemplate) *wfv1.Workflow {
-	wf := toWorkflow(template.TypeMeta, template.ObjectMeta, template.Spec.WorkflowSpec)
-	wf.Labels[LabelKeyClusterWorkflowTemplate] = template.ObjectMeta.Name
+func ConvertWorkflowTemplateToWorkflow(templateName string, clusterScope bool) *wfv1.Workflow {
```
Comment (jessesuen) [inline workflow/common/convert.go:29]: This function should be renamed since "convert" is misleading. I suggest `NewWorkflowFromWorkflowTemplate`

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 7]
```diff
@@ -809,6 +812,22 @@ type TemplateRef struct {
 	ClusterScope bool `json:"clusterScope,omitempty" protobuf:"varint,4,opt,name=clusterScope"`
 }
 
+// WorkflowTemplateRef is a reference of workflow template resource on Top level workflow.
+type WorkflowTemplateRef struct {
+	// Name is the resource name of the workflow template.
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+	// ClusterScope indicates the referred template is cluster scoped (i.e., a ClusterWorkflowTemplate).
```
Comment (jessesuen) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:819]: Remove comma after `i.e.,`

File: workflow/controller/operator.go [hunk 8]
```diff
@@ -87,6 +89,12 @@ type wfOperationCtx struct {
 	// preExecutionNodePhases contains the phases of all the nodes before the current operation. Necessary to infer
 	// changes in phase for metric emission
 	preExecutionNodePhases map[string]wfv1.NodePhase
+	// wfSpec holds the WorkflowSpec for execution
+	wfSpec *wfv1.WorkflowSpec
+	// entrypoint is the starting point for workflow execution
+	entrypoint string
+	// arguments holds all global argument for workflow.
+	arguments wfv1.Arguments
```
Comment (jessesuen) [inline workflow/controller/operator.go:97]: Why did we introduce the `wfSpec`,  as opposed to rewriting `woc.wf.Spec`? The current approach requires everyone to remember *not* to access `wf.Spec` which could be fragile. It would be better if we could avoid two sources of truths.

File: ui/src/models/workflows.ts [hunk 9]
```diff
@@ -766,6 +766,11 @@ export interface WorkflowStatus {
      * Conditions is a list of WorkflowConditions
      */
     conditions?: WorkflowCondition[];
+
+    /**
+     * StoredWorkflowTemplateSpec is a Workflow Spec of top level WorkflowTemplate.
+     */
+    storedWorkflowTemplateSpec?: WorkflowSpec;
```
Comment (simster7) [inline ui/src/models/workflows.ts:773]:
```suggestion
    storedWorkflowSpec?: WorkflowSpec;
```

File: ui/src/models/workflows.ts [hunk 10]
```diff
@@ -766,6 +766,11 @@ export interface WorkflowStatus {
      * Conditions is a list of WorkflowConditions
      */
     conditions?: WorkflowCondition[];
+
+    /**
+     * StoredWorkflowTemplateSpec is a Workflow Spec of top level WorkflowTemplate.
```
Comment (simster7) [inline ui/src/models/workflows.ts:771]: The same comment as the `*_types.go` file
```suggestion
     * StoredWorkflowSpec stores the top level WorkflowTemplate spec
```

File: util/util.go [hunk 11]
```diff
@@ -57,3 +58,19 @@ func WriteTeriminateMessage(message string) {
 		panic(err)
 	}
 }
+
+// Merge the two parameters Slice and overwriting duplicate keys,
+func MergeParameters(params ...[]wfv1.Parameter) []wfv1.Parameter {
+	resultParams := make([]wfv1.Parameter, 0)
```
Comment (simster7) [inline util/util.go:64]: This is clearer, imo
```suggestion
	var resultParams []wfv1.Parameter
```

File: util/util.go [hunk 12]
```diff
@@ -57,3 +58,19 @@ func WriteTeriminateMessage(message string) {
 		panic(err)
 	}
 }
+
+// Merge the two parameters Slice and overwriting duplicate keys,
```
Comment (simster7) [inline util/util.go:62]: Note: this function doesn't actually overwrite duplicate keys... it just takes the first key. Is this what we want?

File: workflow/controller/operator.go [hunk 13]
```diff
@@ -48,7 +50,11 @@ import (
 
 // wfOperationCtx is the context for evaluation and operation of a single workflow
 type wfOperationCtx struct {
+
```
Comment (simster7) [inline workflow/controller/operator.go:53]: Please remove this blank line
```suggestion
```

File: workflow/controller/operator.go [hunk 14]
```diff
@@ -14,6 +14,8 @@ import (
 	"sync"
 	"time"
 
+	"github.com/argoproj/argo/util"
```
Comment (simster7) [inline workflow/controller/operator.go:17]: This import is out of place.

File: workflow/controller/operator.go (hunk 8 above)
Comment (simster7) [reply workflow/controller/operator.go:102]: This comment doesn't seem to have been addressed yet and I agree with @jessesuen's point.

Is there a reason as to why we still use `wfSpec`? Asking in case you guys had an offline discussion.

File: workflow/controller/operator.go [hunk 15]
```diff
@@ -87,6 +93,13 @@ type wfOperationCtx struct {
 	// preExecutionNodePhases contains the phases of all the nodes before the current operation. Necessary to infer
 	// changes in phase for metric emission
 	preExecutionNodePhases map[string]wfv1.NodePhase
+
+	// wfSpec holds the WorkflowSpec for execution
+	wfSpec *wfv1.WorkflowSpec
+	// entrypoint is the starting point for workflow execution
+	entrypoint string
```
Comment (simster7) [inline workflow/controller/operator.go:100]: This field doesn't seem to be actually used anywhere. Why is this needed? If it isn't, can you remove it?

File: workflow/controller/operator.go [hunk 16]
```diff
@@ -141,14 +155,14 @@ func newWorkflowOperationCtx(wf *wfv1.Workflow, wfc *WorkflowController) *wfOper
...
 	}
-
 	return &woc
 }
 
 // operate is the main operator logic of a workflow. It evaluates the current state of the workflow,
 // and its pods and decides how to proceed down the execution path.
 // TODO: an error returned by this method should result in requeuing the workflow to be retried at a
 // later time
+
```
Comment (simster7) [inline workflow/controller/operator.go:165]: Please remove this blank line
```suggestion
```

File: workflow/controller/operator.go [hunk 17]
```diff
@@ -269,16 +292,13 @@ func (woc *wfOperationCtx) operate() {
...
-		return
-	}
+	var node *wfv1.NodeStatus
 
-	node, err := woc.executeTemplate(woc.wf.ObjectMeta.Name, &wfv1.WorkflowStep{Template: woc.wf.Spec.Entrypoint}, tmplCtx, woc.wf.Spec.Arguments, &executeTemplateOpts{})
+	if woc.wf.Spec.WorkflowTemplateRef != nil {
+		node, err = woc.executeTemplate(woc.wf.ObjectMeta.Name, &wfv1.WorkflowStep{TemplateRef: woc.convertWFTmplRefToTmplRef()}, tmplCtx, woc.arguments, &executeTemplateOpts{})
+	} else {
+		node, err = woc.executeTemplate(woc.wf.ObjectMeta.Name, &wfv1.WorkflowStep{Template: woc.wfSpec.Entrypoint}, tmplCtx, woc.arguments, &executeTemplateOpts{})
+	}
```
Comment (simster7) [inline workflow/controller/operator.go:301]: This `if` statement is hard to read and is largely unnecessary. Why not something like:

```go
	tmpl := &wfv1.WorkflowStep{Template: woc.wfSpec.Entrypoint}
	if woc.wf.Spec.WorkflowTemplateRef != nil {
		tmpl.Template = ""
		tmpl.TemplateRef = woc.convertWFTmplRefToTmplRef()
	}
	node, err := woc.executeTemplate(woc.wf.ObjectMeta.Name, tmpl, tmplCtx, woc.wf.Spec.Arguments, &executeTemplateOpts{})
```
It's way cleaner...

File: workflow/controller/operator.go [hunk 18]
```diff
@@ -2624,3 +2646,60 @@ func (woc *wfOperationCtx) includeScriptOutput(nodeName, boundaryID string) (boo
...
+
+	entrypoint := woc.wf.Spec.Entrypoint
+
+	if entrypoint == "" {
+		entrypoint = woc.wfSpec.Entrypoint
+	}
+	return woc.wf.Spec.WorkflowTemplateRef.ConvertTemplateRef(entrypoint)
+}
+
+func (woc *wfOperationCtx) setWorkflowSpecAndEntrypoint() error {
```
Comment (simster7) [inline workflow/controller/operator.go:2678]: I would call this `loadWorkflowSpec` instead

File: workflow/controller/operator.go [hunk 19]
```diff
@@ -87,6 +93,13 @@ type wfOperationCtx struct {
 	// preExecutionNodePhases contains the phases of all the nodes before the current operation. Necessary to infer
 	// changes in phase for metric emission
 	preExecutionNodePhases map[string]wfv1.NodePhase
+
+	// wfSpec holds the WorkflowSpec for execution
+	wfSpec *wfv1.WorkflowSpec
+	// entrypoint is the starting point for workflow execution
+	entrypoint string
+	// arguments holds all global argument for workflow.
+	arguments wfv1.Arguments
```
Comment (simster7) [inline workflow/controller/operator.go:102]: Also, it seems that whenever this field is used, only the parameters are accessed. I would change this to
```suggestion
	submissionParameters []wfv1.Parameter
```

File: workflow/controller/operator.go [hunk 20]
```diff
@@ -2624,3 +2646,60 @@ func (woc *wfOperationCtx) includeScriptOutput(nodeName, boundaryID string) (boo
...
+	return woc.wf.Spec.WorkflowTemplateRef.ConvertTemplateRef(entrypoint)
+}
+
+func (woc *wfOperationCtx) setWorkflowSpecAndEntrypoint() error {
+
+	woc.arguments.Parameters = woc.wf.Spec.Arguments.Parameters
+	woc.entrypoint = woc.wf.Spec.Entrypoint
+
+	if woc.wf.Spec.WorkflowTemplateRef == nil {
+		woc.wfSpec = woc.wf.Spec.DeepCopy()
```
Comment (simster7) [inline workflow/controller/operator.go:2684]: `DeepCopy` is an expensive call, what's wrong with `woc.wfSpec = woc.wf.Spec`?

File: workflow/controller/operator.go [hunk 21]
```diff
@@ -2624,3 +2646,60 @@ func (woc *wfOperationCtx) includeScriptOutput(nodeName, boundaryID string) (boo
...
+		return nil
+	}
+	if woc.wf.Status.StoredWorkflowTemplateSpec != nil {
```
Comment (simster7) [inline workflow/controller/operator.go:2687]: This should be an `else if`

File: workflow/controller/operator.go [hunk 22]
```diff
@@ -2624,3 +2646,60 @@ func (woc *wfOperationCtx) includeScriptOutput(nodeName, boundaryID string) (boo
 	}
 	return false, nil
 }
+
+func (woc *wfOperationCtx) getTopLevelWorkflowTemplate() (*wfv1.WorkflowSpec, error) {
```
Comment (simster7) [inline workflow/controller/operator.go:2650]: Don't use the term `TopLevel` here. I would call this
```suggestion
func (woc *wfOperationCtx) fetchWorkflowSpec() (*wfv1.WorkflowSpec, error) {
```

Comment (simster7) [reply workflow/controller/operator.go:2650]: Furthermore, I think this should be a utility function, not a member of `wfOperationCtx`. You can pass in `woc.wf.Spec.WorkflowTemplateRef` as a parameter

File: workflow/controller/operator.go [hunk 23]
```diff
@@ -2624,3 +2646,60 @@ func (woc *wfOperationCtx) includeScriptOutput(nodeName, boundaryID string) (boo
...
+
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+	return wftmpl.GetSpec().DeepCopy(), nil
+}
+
+func (woc *wfOperationCtx) convertWFTmplRefToTmplRef() *wfv1.TemplateRef {
```
Comment (simster7) [inline workflow/controller/operator.go:2668]: This function is only used once, so I would just add the code in-place where it is used.

File: workflow/controller/operator_metrics_test.go [hunk 24]
```diff
@@ -50,6 +50,8 @@ func TestBasicMetric(t *testing.T) {
 	_, err := wfcset.Create(wf)
 	assert.NoError(t, err)
 	woc := newWorkflowOperationCtx(wf, controller)
+	err = woc.setWorkflowSpecAndEntrypoint()
```
Comment (simster7) [inline workflow/controller/operator_metrics_test.go:53]: If you have to add this function anytime a `WorkflowOperationCtx` then it's probably implemented incorrectly. Setting the `WorkflowSpec` and `Entrypoint` is an implementation detail of `WorkflowOperationCtx` and should not be exposed externally.

Is there a reason as to why you do this? If not, please refactor so that making these calls externally is not necessary.

File: workflow/validate/validate.go [hunk 25]
```diff
@@ -104,24 +105,59 @@ func ValidateWorkflow(wftmplGetter templateresolution.WorkflowTemplateNamespaced
...
+	hasWorkflowTemplateRef := wf.Spec.WorkflowTemplateRef != nil
+
+	if hasWorkflowTemplateRef {
+		err := ValidateFieldsWithWFTRef(wf.Spec)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if hasWorkflowTemplateRef {
```
Comment (simster7) [inline workflow/validate/validate.go:122]: Why are there two identical `if hasWorkflowTemplateRef {` one right after the other? Should you join them together?

File: workflow/validate/validate.go [hunk 26]
```diff
@@ -104,24 +105,59 @@ func ValidateWorkflow(wftmplGetter templateresolution.WorkflowTemplateNamespaced
 	ctx := newTemplateValidationCtx(wf, opts)
 	tmplCtx := templateresolution.NewContext(wftmplGetter, cwftmplGetter, wf, wf)
 
-	err := validateWorkflowFieldNames(wf.Spec.Templates)
+	var wftmpl wfv1.WorkflowSpecHolder
+	var entrypoint = wf.Spec.Entrypoint
```
Comment (simster7) [inline workflow/validate/validate.go:109]:
```suggestion
	entrypoint := wf.Spec.Entrypoint
```

File: workflow/validate/validate.go [hunk 27]
```diff
@@ -104,24 +105,59 @@ func ValidateWorkflow(wftmplGetter templateresolution.WorkflowTemplateNamespaced
...
+	var topLevelTmplRef *wfv1.TemplateRef
```
Comment (simster7) [inline workflow/validate/validate.go:110]: Please don't use the term "top level"

File: workflow/validate/validate.go [hunk 28]
```diff
@@ -104,24 +105,59 @@ func ValidateWorkflow(wftmplGetter templateresolution.WorkflowTemplateNamespaced
 	ctx := newTemplateValidationCtx(wf, opts)
 	tmplCtx := templateresolution.NewContext(wftmplGetter, cwftmplGetter, wf, wf)
 
-	err := validateWorkflowFieldNames(wf.Spec.Templates)
+	var wftmpl wfv1.WorkflowSpecHolder
```
Comment (simster7) [inline workflow/validate/validate.go:108]:
```suggestion
	var wfSpecHolder wfv1.WorkflowSpecHolder
```

File: workflow/validate/validate.go [hunk 29]
```diff
@@ -214,15 +254,42 @@ func ValidateWorkflow(wftmplGetter templateresolution.WorkflowTemplateNamespaced
 	return wfConditions, nil
 }
 
+func ValidateFieldsWithWFTRef(v interface{}) error {
+	val := reflect.ValueOf(v)
+	// Acceptable fields if workflow has top level workflow template reference.
```
Comment (simster7) [inline workflow/validate/validate.go:259]: Please don't use the term "top level"

File: workflow/validate/validate.go [hunk 30]
```diff
@@ -214,15 +254,42 @@ func ValidateWorkflow(wftmplGetter templateresolution.WorkflowTemplateNamespaced
 	return wfConditions, nil
 }
 
+func ValidateFieldsWithWFTRef(v interface{}) error {
```
Comment (simster7) [inline workflow/validate/validate.go:257]: Please choose a different name for this, perhaps?
```suggestion
func ValidateWorkflowSpecFields(v interface{}) error {
```

File: workflow/validate/validate.go [hunk 31]
```diff
@@ -214,15 +254,42 @@ func ValidateWorkflow(wftmplGetter templateresolution.WorkflowTemplateNamespaced
...
+	val := reflect.ValueOf(v)
```
Comment (simster7) [inline workflow/validate/validate.go:258]: Why do we need to use `reflect` here at all if the only time this function is called is with a `WorkflowSpec` type?

This method is not only hard to read, but it's also susceptible to runtime errors and is more expensive to run.

File: workflow/validate/validate.go [hunk 32]
```diff
@@ -214,15 +254,42 @@ func ValidateWorkflow(wftmplGetter templateresolution.WorkflowTemplateNamespaced
...
+		"Arguments":             true,
+		"WorkflowTemplateRef":   true,
+	}
+	for i := 0; i < val.NumField(); i++ {
+		// Lookup the validate tag
+		field := val.Type().Field(i)
+		fieldVal := val.Field(i).IsZero()
+		_, ok := validFields[field.Name]
+		if !ok && !fieldVal {
+			return errors.Errorf(errors.CodeBadRequest, "%s is invalid field in spec if workflow has top level template reference", field.Name)
```
Comment (simster7) [inline workflow/validate/validate.go:274]: Please don't use the term "top level"

File: pkg/apis/workflow/v1alpha1/common.go [hunk 33]
```diff
@@ -22,6 +22,10 @@ type TemplateHolder interface {
 	GetResourceScope() ResourceScope
 }
 
+type WorkflowSpecHolder interface {
+	GetSpec() *WorkflowSpec
```
[the remark this answers is not in the document]
Comment (sarabala1979) [reply pkg/apis/workflow/v1alpha1/common.go]: I think no need to change Cronworkflow. `WorkflowSpec` should have `WorkflowTempalteRef`.
```
apiVersion: argoproj.io/v1alpha1
kind: CronWorkflow
metadata:
  name: hello-world
spec:
  schedule: "* * * * *"
  timezone: "America/Los_Angeles"   # Default to local machine timezone
  startingDeadlineSeconds: 0
  concurrencyPolicy: "Replace"      # Default to "Allow"
  successfulJobsHistoryLimit: 4     # Default 3
  failedJobsHistoryLimit: 4         # Default 1
  suspend: false                    # Set to "true" to suspend scheduling
  workflowSpec:
    entrypoint: whalesay
    workflowTemplateRef:
        name: workflow-template-submittable
```

File: workflow/controller/operator.go (hunk 8 above)
Comment (sarabala1979) [reply workflow/controller/operator.go:102]: Discussed with @jessesuen. he is ok with wfSpec. we need to maintain the two spec in this scenario. One is for execution which comes from WFT and updates workflow properties like suspend, activedeadline, and arguments.

File: workflow/validate/validate.go (hunk 31 above)
Comment (sarabala1979) [reply workflow/validate/validate.go:268]: Is there any other recommendation to validate the fields in a struct?  I tried with Json marshal and Structs to Map options But WorkflowSpec Struct has few primitive data types. it didn't work.

File: util/util.go (hunk 12 above)
Comment (sarabala1979) [reply util/util.go:62]: yes.  passing map in order, the first one take priority

File: workflow/controller/operator.go (hunk 15 above)
Comment (sarabala1979) [reply workflow/controller/operator.go:100]: Good catch

File: workflow/controller/operator.go (hunk 21 above)
Comment (sarabala1979) [reply workflow/controller/operator.go]: there is no statement after if statement for the first condition that's why there is a return statement inside block. I think `else if` is not needed

File: workflow/controller/operator.go (hunk 22 above)
Comment (sarabala1979) [reply workflow/controller/operator.go]: I thought too. But function needs to access the informer from wfOperationCtx. I thought instead of passing all references. This function is going used only in WorkflowTemplate usecase. it would be better to have it simple and clean.

File: util/util.go (hunk 12 above)
Comment (simster7) [reply util/util.go:62]: Got it. Can we update this comment to reflect that there is no overwriting being done?

File: workflow/controller/operator.go [hunk 34]
```diff
@@ -132,6 +142,7 @@ func newWorkflowOperationCtx(wf *wfv1.Workflow, wfc *WorkflowController) *wfOper
 		deadline:               time.Now().UTC().Add(maxOperationTime),
 		auditLogger:            argo.NewAuditLogger(wf.ObjectMeta.Namespace, wfc.kubeclientset, wf.ObjectMeta.Name),
 		preExecutionNodePhases: make(map[string]wfv1.NodePhase),
+		submissionParameters:   wf.Spec.DeepCopy().Arguments.Parameters,
```
Comment (simster7) [inline workflow/controller/operator.go:145]: Should this be
```suggestion
		submissionParameters:   wf.Spec.Arguments.Parameters.DeepCopy(),
```

Instead?

File: workflow/controller/operator.go [hunk 35]
```diff
@@ -269,16 +288,23 @@ func (woc *wfOperationCtx) operate() {
...
 
-	// Create a starting template context.
-	tmplCtx, err := woc.createTemplateContext(wfv1.ResourceScopeLocal, "")
-	if err != nil {
-		msg := fmt.Sprintf("Failed to create a template context: %+v", err)
-		woc.log.Errorf(msg)
-		woc.markWorkflowError(err, true)
-		return
+	var node *wfv1.NodeStatus
+
```
Comment (simster7) [inline workflow/controller/operator.go:292]: Much better! You can remove this blank line here
```suggestion
```

File: workflow/controller/operator.go [hunk 36]
```diff
@@ -2651,3 +2679,48 @@ func (woc *wfOperationCtx) includeScriptOutput(nodeName, boundaryID string) (boo
 	}
 	return false, nil
 }
+
+func (woc *wfOperationCtx) fetchWorkflowTemplate() (*wfv1.WorkflowSpec, error) {
```
Comment (simster7) [inline workflow/controller/operator.go:2683]:
```suggestion
func (woc *wfOperationCtx) fetchWorkflowSpec() (*wfv1.WorkflowSpec, error) {
```

This function actually returns a `WorkflowSpec`, not a `WorkflowTemplate`, so it should be named appropriately.

File: workflow/controller/operator_metrics_test.go (hunk 24 above)
Comment (sarabala1979) [reply workflow/controller/operator_metrics_test.go]: In Test, Controller function processNextItem will not be called. That's why we need to make call for `operate()` and this function on all test functions to mimic the real call flow.

Comment (sarabala1979) [reply workflow/controller/operator_metrics_test.go]: moved `loadWorkflowSpec` inside operate function

File: Makefile [hunk 37]
```diff
@@ -478,8 +478,8 @@ pkg/apis/workflow/v1alpha1/openapi_generated.go:
 pkg/apiclient/_.secondary.swagger.json: hack/secondaryswaggergen.go pkg/apis/workflow/v1alpha1/openapi_generated.go dist/kubernetes.swagger.json
 	go run ./hack secondaryswaggergen
 
-dist/swagger.json: $(GOPATH)/bin/swagger $(SWAGGER_FILES) $(MANIFESTS_VERSION_FILE) hack/swaggify.sh
-	swagger mixin -c 680 $(SWAGGER_FILES) | sed 's/VERSION/$(MANIFESTS_VERSION)/' | ./hack/swaggify.sh > dist/swagger.json
+dist/swagger.json: $(HOME)/go/bin/swagger $(SWAGGER_FILES) $(MANIFESTS_VERSION_FILE) hack/swaggify.sh
```
Comment (alexec) [inline Makefile:481]: you must revert this to $(GOPATH)/bin/swagger

File: workflow/controller/operator.go [hunk 38]
```diff
@@ -169,6 +179,15 @@ func (woc *wfOperationCtx) operate() {
 
 	woc.log.Infof("Processing workflow")
 
+	// Loading WorkflowSpec for execution.
+	err := woc.loadWorkflowSpec()
+	if err != nil {
+		woc.log.Errorf("Unable to get Workflow Template Reference for workflow, %s error: %s", woc.wf.Name, err)
+		woc.markWorkflowError(err, true)
+		woc.persistUpdates()
```
Comment (simster7) [inline workflow/controller/operator.go:187]: `persistUpdates()` is called as a `defer` function above, so by adding it here you'll be calling it twice and making two API calls to `Update`. This should be removed.
```suggestion
```

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 39]
```diff
@@ -1742,6 +1764,11 @@ func (wf *Workflow) GetResourceScope() ResourceScope {
 	return ResourceScopeLocal
 }
 
+// GetWorkflowSpec returns the Spec of a workflow.
+func (wf *Workflow) GetSpec() WorkflowSpec {
```
Comment (simster7) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:1768]: Shouldn't this be called
```suggestion
func (wf *Workflow) GetWorkflowSpec() WorkflowSpec {
```
? The comment above is correct but the name is not.

File: workflow/controller/controller.go [hunk 40]
```diff
@@ -447,15 +447,17 @@ func (wfc *WorkflowController) processNextItem() bool {
 		wfc.throttler.Remove(key)
 		return true
 	}
+
 	startTime := time.Now()
+
```
Comment (simster7) [inline workflow/controller/controller.go:452]: Please delete this blank line
```suggestion
```

File: workflow/controller/operator.go [hunk 41]
```diff
@@ -269,16 +297,22 @@ func (woc *wfOperationCtx) operate() {
...
+			entrypoint = woc.wfSpec.Entrypoint
+		}
+		tmpl.Template = ""
+		tmpl.TemplateRef = woc.wf.Spec.WorkflowTemplateRef.ToTemplateRef(entrypoint)
+		args.Parameters = woc.submissionParameters
 	}
 
-	node, err := woc.executeTemplate(woc.wf.ObjectMeta.Name, &wfv1.WorkflowStep{Template: woc.wf.Spec.Entrypoint}, tmplCtx, woc.wf.Spec.Arguments, &executeTemplateOpts{})
+	node, err = woc.executeTemplate(woc.wf.ObjectMeta.Name, tmpl, tmplCtx, args, &executeTemplateOpts{})
+
```
Comment (simster7) [inline workflow/controller/operator.go:315]: Please delete this blank line
```suggestion
```

File: workflow/controller/operator.go [hunk 42]
```diff
@@ -2651,3 +2687,48 @@ func (woc *wfOperationCtx) includeScriptOutput(nodeName, boundaryID string) (boo
...
+func (woc *wfOperationCtx) fetchWorkflowSpec() (*wfv1.WorkflowSpec, error) {
+	var wftmpl wfv1.WorkflowSpecHolder
+	var err error
+	// Logic for workflow refers Workflow template
+	if woc.wf.Spec.WorkflowTemplateRef != nil {
+		if woc.wf.Spec.WorkflowTemplateRef.ClusterScope {
+			wftmpl, err = woc.controller.cwftmplInformer.Lister().Get(woc.wf.Spec.WorkflowTemplateRef.Name)
+		} else {
+			wftmpl, err = woc.controller.wftmplInformer.Lister().WorkflowTemplates(woc.wf.Namespace).Get(woc.wf.Spec.WorkflowTemplateRef.Name)
+
```
Comment (simster7) [inline workflow/controller/operator.go:2700]: Please delete this blank line
```suggestion
```

File: workflow/controller/operator.go [hunk 43]
```diff
@@ -2651,3 +2687,48 @@ func (woc *wfOperationCtx) includeScriptOutput(nodeName, boundaryID string) (boo
...
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+	return wftmpl.GetWorkflowSpec().DeepCopy(), nil
+}
+
+func (woc *wfOperationCtx) loadWorkflowSpec() error {
+
```
Comment (simster7) [inline workflow/controller/operator.go:2710]: Please delete this blank line
```suggestion
```

File: workflow/controller/operator_test.go [hunk 44]
```diff
@@ -1826,8 +1831,10 @@ func TestAddGlobalArtifactToScope(t *testing.T) {
 func TestParamSubstitutionWithArtifact(t *testing.T) {
 	wf := test.LoadE2EWorkflow("functional/param-sub-with-artifacts.yaml")
 	woc := newWoc(*wf)
+	err := woc.loadWorkflowSpec()
```
Comment (simster7) [inline workflow/controller/operator_test.go:1834]: This is now present on `woc.opearte()` should it be removed here?

File: workflow/controller/operator_top_wftmpl_ref_test.go [hunk 45]
```diff
@@ -0,0 +1,135 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	wfv1 "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+	"github.com/argoproj/argo/util"
+)
+
+func TestTopLevelWFTmplRef(t *testing.T) {
+	//_, controller := newController()
```
Comment (simster7) [inline workflow/controller/operator_top_wftmpl_ref_test.go:13]: Delete this comment?

File: pkg/apis/workflow/v1alpha1/workflow_types.go (hunk 39 above)
Comment (simster7) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:1768]: Otherwise this won't be counted as an interface implementation

**Comment by dtaniwaki** [comment]
Thank you for implementing this feature 🆒
